src/Oscoin/Crypto/Blockchain/Eval.hs

```diff
@@ -179,39 +169,39 @@ evalToState eval tx = state go
 -- | Return an unsealed block holding the given transactions and state
 -- on top of the given parent block.
 mkUnsealedBlock
-    :: forall c t tx st.
-       ( Foldable t
-       , Serialise tx
+    :: forall c tx s st.
+       ( Serialise tx
        , AuthTree.MerkleHash (Hash c)
        , Crypto.Hashable c (BlockHeader c Unsealed)
        , Crypto.Hashable c st
        )
-    => BlockHash c -> Timestamp -> t tx -> st -> Block c tx Unsealed
+    => Maybe (Block c tx s)
+    -> Timestamp
+    -> [tx]
+    -> st
+    -> Block c tx Unsealed
 mkUnsealedBlock parent blockTimestamp txs blockState =
     let (hdr :: BlockHeader c Unsealed) = emptyHeader
+          { blockDataHash  = hashTxs txs
+          , blockStateHash = Crypto.fromHashed (Crypto.hash blockState)
+          , blockTimestamp
+          }
     in mkBlock (header hdr) txs
   where
-    header hdr = hdr
-        { blockPrevHash  = parent
-        , blockDataHash  = hashTxs txs
-        , blockStateHash = Crypto.fromHashed (Crypto.hash blockState)
-        , blockTimestamp
-        }
+    header hdr = case parent of
+        Nothing -> hdr -- We are creating 'genesis'
+        Just p  -> hdr { blockHeight    = succ . blockHeight . blockHeader $ p
+                       , blockPrevHash  = blockHash p
+                       }
 
 mkEvaledUnsealedBlock
-    :: ( Serialise tx
-       , Crypto.Hashable c st
-       , Crypto.Hashable c (BlockHeader c Unsealed)
-       , AuthTree.MerkleHash (Hash c)
-       )
-    => Evaluator st tx o
-    -> Timestamp
+    :: (st -> Block c tx Unsealed)
     -> [tx]
-    -> BlockHash c
+    -> Evaluator st tx o
     -> st
     -> Either (tx, EvalError) (Block c tx Unsealed)
-mkEvaledUnsealedBlock eval tick txs parent initState =
-    mkUnsealedBlock parent tick txs <$> foldlM step initState txs
+mkEvaledUnsealedBlock createBlock txs eval initState =
+    createBlock <$> foldlM step initState txs
```

Comment:
How about calling turning this into `evalEither txs initState = foldM step initState txs` and applying `createBlock` explicitly at the call site instead of requiring a callback?

Reply:
Works for me (despite I don't necessarily see anything evil with the callback approach given all these functions are internal), as the intended purpose here was simply to alleviate the impedance mismatch between creating a "normal" unsealed block vs genesis, where we don't have a parent block at hand 😉

Reply:
I just found the type signature hard to understand.

Reply:
Done! 😉 04fc7743